AI console info, move declaration and alpha beta pruning

PinMeister opened the pull request:
- Add and track played pieces for AI
- Implemented alpha-beta pruning
- Implemented move declaration (even moving to or from X10)
- AI declares if they are placing or moving a piece at the specified position
- AI can only place tokens if they have tokens remaining
- AI only considers moves among all the played pieces they have
- Make debug mode to play with AI
- Added coin toss mechanic to determine if player or AI move first (which also determines if AI is maximizing player or not)
- Added timer to show how long AI turn takes
- DID NOT touch the heuristic itself

PinMeister committed:
Print additional information for AI turns

PinMeister committed:
Merge branch 'master' into console-info

PinMeister committed:
AI can move pieces, but tries to move opponent's pieces

PinMeister committed:
AI moves only played pieces, debug mode with AI

PinMeister committed:
Updated move declaration to reflect tournament requirements

PinMeister committed:
Alpha beta pruning implemented

PinMeister committed:
AI can go either first or second

PinMeister committed:
AI can only play tokens if they have tokens remaining

PinMeister committed:
Coin toss fix

Songtone commented on Game.py:
```diff
@@ -85,40 +93,82 @@ def play(self):
             player2_turn += 1
 
     def play_with_ai(self):
+        coin_toss = random.randint(1, 2)
```
I believe its not a coin toss. I think the TA wants to be able to choose who goes first. So it's not necessary to have the coin toss.

Songtone reviewed (changes requested):
Nicely Done! Just a small change

JamalG16 reviewed (changes requested):
![image](https://user-images.githubusercontent.com/35748873/69640786-fac45e00-102c-11ea-96b9-db630e9f8078.png)
This kind of behavior didn't happen before alpha-beta, the AI just put 3 pieces vertically after I blocked it. It doesn't seem to be functioning properly

It's placing tokens almost randomly it seems, something is not working with the algorithm
![image](https://user-images.githubusercontent.com/35748873/69641084-89d17600-102d-11ea-95cc-7107a449283c.png)

PinMeister committed:
Manual coin toss

PinMeister committed:
Alpha beta bug fix

PinMeister commented:
I set max_depth to 0 and made it so that every time alpha beta is called, depth decreases by 1 rather than increase by 1. This should more accurately reflect alpha beta pruning.

JamalG16 commented:
![image](https://user-images.githubusercontent.com/35748873/69660168-b6958580-104d-11ea-8902-9fc3bc7fa267.png) with depth 0

any depth greater than 0 or 1 the program comes to a standstill and doesnt progress
![image](https://user-images.githubusercontent.com/35748873/69660214-d2009080-104d-11ea-9f7c-2be387ac503b.png)

with depth 1:
![image](https://user-images.githubusercontent.com/35748873/69660281-f3fa1300-104d-11ea-9ec6-8e503a9ecb46.png)


this is all behavior that has changed since alpha-beta pruning, with the previous version of minimax without alpha beta pruning the AI would have behaved differently so I know it's not the heuristic that's causing these issues, something is wrong with your aglorithm


Please test your aglorithm by playing with it and then playing with the version pre-alpha beta to ensure that the behavior is the same before pushing fixes

PinMeister committed:
Revert alpha beta back to minimax

Songtone committed:
Added pruning

PinMeister committed:
Fix player movement from X10

PinMeister committed:
New heuristic

- If AI goes first, it tries to make a X
- If AI goes second, it actively tries to block the player

PinMeister committed:
Force draw if both player and AI have no more tokens and moves

PinMeister committed:
Stronger blocking heuristic
